rules/aip0123/resource_pattern_singular.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package aip0123 contains rules defined in https://aip.dev/123.
+package aip0123
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/googleapis/api-linter/lint"
+	"github.com/googleapis/api-linter/locations"
+	"github.com/googleapis/api-linter/rules/internal/utils"
+	"github.com/jhump/protoreflect/desc"
+	"github.com/stoewer/go-strcase"
+)
+
+var resourcePatternSingular = &lint.MessageRule{
+	Name: lint.NewRuleName(123, "resource-pattern-singular"),
+	OnlyIf: func(m *desc.MessageDescriptor) bool {
+		return utils.IsResource(m) && len(utils.GetResource(m).GetPattern()) > 0
+	},
+	LintMessage: func(m *desc.MessageDescriptor) []lint.Problem {
+		var problems []lint.Problem
+		res := utils.GetResource(m)
+		nested := isNestedName(res)
+		var nn string
+		if nested {
+			nn = nestedSingular(res)
+		}
+
+		patterns := res.GetPattern()
+		singular := utils.GetResourceSingular(res)
+
+		if !utils.IsSingletonResource(m) {
+			singular = fmt.Sprintf("{%s}", strcase.SnakeCase(singular))
+			nn = fmt.Sprintf("{%s}", nn)
```

Comment:
If you have a non-nested singleton will this be empty as `nn` will be unset? Or is that case impossible?

Reply:
I don't think non-nested singletons are possible since they depend on the lifecycle of another resource e.g. when X is created, XSingleton is also created. I don't know of any examples of them either.

That said, this specific line wouldn't be executed b.c the check "if no singleton" above would skip this in that case.

but having top-level singletons shouldn't be a thing...this did remind me though that I needed to better handle that case in the "is root level" check (fixed).